text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,217 @@
+- Feature Name: `gem_checksum_verification`
+- Start Date: 2023-08-29
+- RFC PR: https://github.com/rubygems/rubygems/pull/6374
+- Bundler Issue: (leave this empty)
+
+# Summary
+
+Bundler is adding checksum verification of gems when they are installed. It should be secure by default and easy to use. It should not break assumptions or unnecessarily break deployment or CI unless there is a security problem.
+
+# Motivation
+
+Verifying gem checksums with known checksums at install time is a stronger way to verify that the exact same gem source is being used in every environment where the bundle is installed.
+Using checksums, sourced from rubygems.org or from digests of local .gem file, is a more strict way of locking to a specific gem version.
+The feature should work transparently in much the same way that bundler already locks to specific versions by ensuring that not just the same version, but the exact same rubygem file data, is installed on each environment.
+
+# Guide-level explanation
+
+Upon first upgrading to this version of Bundler, expect bundler to continue to be secure by default and to progressively add checksums to the lockfile.
+
+If you wish to immediately add checksums to your lockfile for all locked gems, run `bundle lock`.
+Bundle lock now fetches checksums from remote sources by default.
+If you would like to bypass this behavior, add the `--no-checksums` flag.
+
+Common Bundler commands like `bundle install`, `bundle update`, `bundle lock`, `bundle add` will now automatically record and verify checksums.
+Commands that rely on checksums for verification will output a message or warning when checksums are missing or mismatched.
+
+Example:
+
+```
+$ bundle install
+Bundle complete! 88 Gemfile dependencies, 256 gems now installed.
+Use `bundle info [gemname]` to see where a bundled gem is installed.
+Use `bundle lock` to add missing checksums to Gemfile.lock.
+```
+
+Running `bundle update` or `bundle add` will record the checksum from the source (e.g. rubygems.org) into the Gemfile.lock, if it is available.
+If a checksum is not available from the source because the source does not provide such info (e.g. private gemservers) then a checksum will be created during install using the .gem file.
+If a checksum can't be created because the source is a path or git source, then only the gems name and version will be recorded with a blank checksum.
+
+If you want to ensure that the bundled environment contains only gems matching the checksums in the lockfile, run `bundle pristine`.
+The `bundle pristine` command installs every gem fresh from a newly downloaded .gem source.
+The pristine install will trigger computation and comparison of the generated SHA256 checksum with the checksum stored in the lockfile.
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems without checksums.
+Use `bundle lock` to add checksums to Gemfile.lock.
```

Comment:
I originally wrote this because it seemed like a nice interface, but now I wonder how I can reliably tell how many gems _could_ have checksums that don't.

Is this useful and we should add it?

Reply:
This also brings up the problem of not continually printing this when the source doesn't support checksums. If the source doesn't support it, do we add a special checksum like `unavailable=true` and then use that as a marker for a gem that doesn't need to be verified? We do support just about any key value pair in the checksum field.

Reply:
When I first saw this, I understood `bundle install` by default would only advertise checksums, but not add them, and I thought that was ideal, regardless of showing a number or not. I.e., let people know that this feature is available and tell them how to enable it. And don't change behavior unless a CHECKSUM section is present in the lockfile.

Reply:
Given that we really can't anticipate the problems of this feature, I'm starting to lean this way.

So the plan would be to build into this feature a line that suggests running `bundle lock --checksums` at the end of every install message until the CHECKSUMS exist in the Gemfile.lock. Until then, we treat the feature as disabled.

Reply:
Yes! I think a line in `bundle install` output is not too invasive and properly advertises the feature.